WordPress/src/main/java/org/wordpress/android/ui/posts/PostEditorAnalyticsSession.java

```diff
@@ -0,0 +1,132 @@
+    private String mPostType;
+    private String mBlogType;
+    private String mContentType;
+    private boolean mStarted = false;
+    private Editor mCurrentEditor;
+    private boolean mHasUnsupportedBlocks = false;
+    private Outcome mOutcome = null;
+
+    enum Editor {
+        GUTENBERG,
+        CLASSIC,
+        HTML
+    }
+
+    enum Outcome {
+        CANCEL,
+        DISCARD,    // not used in WPAndroid, but kept for parity with iOS
+                    // see https://github.com/wordpress-mobile/gutenberg-mobile/issues/556#issuecomment-462678807
+        SAVE,
+        PUBLISH
+    }
+
+    PostEditorAnalyticsSession(Editor editor, PostModel post, SiteModel site) {
+        // fill in which the current Editor is
+        mCurrentEditor = editor;
+
+        // fill in mPostType
+        if (post.isPage()) {
+            mPostType = "page";
+        } else {
+            mPostType = "post";
+        }
+
+        // fill in mBlogType
+        if (site.isWPCom()) {
+            mBlogType = "wpcom";
+        } else if (site.isJetpackConnected()) {
+            mBlogType = "jetpack";
+        } else {
+            mBlogType = "core";
+        }
+
+        // fill in mContentType
+        String postContent = post.getContent();
+        if (TextUtils.isEmpty(post.getContent())) {
+            mContentType = "new";
+        } else if (PostUtils.contentContainsGutenbergBlocks(postContent)) {
+            mContentType = "gutenberg";
+        } else {
+            mContentType = "classic";
+        }
+    }
+
+    public void start(boolean hasUnsupportedBlocks) {
+        if (!mStarted) {
+            mHasUnsupportedBlocks = hasUnsupportedBlocks;
+            Map<String, Object> properties = getCommonProperties();
+            AnalyticsTracker.track(Stat.EDITOR_SESSION_START, properties);
+            mStarted = true;
+        } else {
+            AppLog.w(T.EDITOR, "An editor session cannot be started more than once, "
+                               + "unless it's due to rotation or Editor switch");
+        }
+    }
+
+    public void switchEditor(Editor editor) {
+        mCurrentEditor = editor;
+        Map<String, Object> properties = getCommonProperties();
+        AnalyticsTracker.track(Stat.EDITOR_SESSION_SWITCH_EDITOR, properties);
+    }
+
+    public void setOutcome(Outcome newOutcome) {
+        // on WPiOS we're using `forceOutcome` and some exception logic to allow / not allow forcing an outcome,
+        // but in WPAndroid we slightly changed the interface as we're not
+        // really `forcing` but just `setting` the outcome on specific places where we know what happened
+        // or is going to happen.
+        // Session ending will only properly end the session with the outcome already being set.
+        mOutcome = newOutcome;
+    }
+
+    public void end() {
+        // don't try to send an "end" event if the session wasn't started in the first place
+        if (mStarted) {
```

Comment:
Calling `end()` if not started yet is an error condition? If yes, wdyt about logging a message if not started?

Reply:
addressed in 2982188834, also clarified log comment for `start()`  in fb5972b20a